tests/devices/test_default_qubit.py

```diff
@@ -1989,3 +1661,133 @@ def test_s_inverse():
 
         expected = np.array([1., -1.j]) / np.sqrt(2)
         assert np.allclose(dev.state, expected, atol=tol, rtol=0)
+
+@pytest.mark.parametrize("inverse", [True, False])
+class TestApplyOperationUnit:
+    """Unit tests for the internal _apply_operation method."""
+
+    def test_internal_apply_ops_case(self, inverse):
+        """Tests that if we provide an operation that has an internal
+        implementation, then we use that specific implementation."""
+        dev = qml.device('default.qubit', wires=1)
+
+        # Create a dummy operation
+        expected_test_output = np.ones(1)
+        paulix = lambda *args, **kwargs: expected_test_output
+
+        # Set the internal ops implementations dict
+        dev._apply_ops = {"PauliX": paulix}
+
+        test_state = np.array([1,0])
+        op = qml.PauliX(0) if not inverse else qml.PauliX(0).inv()
+
+        res = dev._apply_operation(test_state, op)
+        assert np.allclose(res, expected_test_output)
+
+    def test_diagonal_operation_case(self, inverse, mocker):
+        """Tests the case of applying a diagonal operation."""
```

Comment:
Docstring could potentially be made more specific

Reply:
Updated